zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,755 @@
+ *
+ * @retval 0 on success;
+ * @retval -ENOTEMPTY on heap having active allocations.
+ */
+int vmh_free(struct vmh_heap *heap, void *ptr)
+{
+	int retval;
+
+	if (heap->core_id != cpu_get_id())
+		return -EINVAL;
+
+	size_t mem_block_iterator, i, size_to_free, block_size, ptr_bit_array_offset,
+		ptr_bit_array_position, physical_block_count, phys_block_end,
+		check_offset, check_position, check_size;
+	uintptr_t phys_aligned_ptr, phys_aligned_alloc_end, phys_block_ptr,
+		block_realigned_ptr, block_realigned_alloc_end;
+	bool ptr_range_found;
+
+	/* Get allocator from which ptr was allocated */
+	for (mem_block_iterator = 0, ptr_range_found = false;
+		mem_block_iterator < MAX_MEMORY_ALLOCATORS_COUNT;
+		mem_block_iterator++) {
+		block_size =
+			1 << heap->physical_blocks_allocators[mem_block_iterator]->blk_sz_shift;
+
+		if (vmh_is_ptr_in_memory_range(ptr,
+				(void *)heap->physical_blocks_allocators
+				[mem_block_iterator]->buffer,
+				heap->physical_blocks_allocators
+				[mem_block_iterator]->num_blocks * block_size)) {
+			ptr_range_found = true;
 			break;
-		default:
+		}
+	}
+	if (!ptr_range_found)
+		return -EINVAL;
+
+	/* Initially set size to block size */
+	size_to_free = block_size;
+
+	/* Get size of the allocation from ptr using allocation_sizes bit array
+	 * and allocators bit array. Only span alloc can have different size than
+	 * block_size.
+	 */
+	if (heap->allocating_continuously) {
+
+		/* Not sure if that is fastest way to find the size comments welcome */
+		ptr_bit_array_offset = (uintptr_t)ptr
+			- (uintptr_t)heap->physical_blocks_allocators[mem_block_iterator]->buffer;
+		ptr_bit_array_position = ptr_bit_array_offset / block_size;
+
+		/* Allocation bit array check */
+		int bit_value, prev_bit_value = 0;
+
+		sys_bitarray_test_bit(heap->allocation_sizes[mem_block_iterator],
+			ptr_bit_array_position, &bit_value);
+		/* If checked bit is in position 0 we assume it is valid
+		 * and assigned 0 for further logic
+		 */
+		if (ptr_bit_array_position)
+			sys_bitarray_test_bit(heap->allocation_sizes[mem_block_iterator],
+				ptr_bit_array_position - 1, &prev_bit_value);
+
+		/* If bit is 1 we know we could be at the start of the allocation,
+		 * we need to test previous bit from allocation if it is set then
+		 * something went wrong in calculations.
+		 */
+		if (prev_bit_value)
 			return -EINVAL;
+
+		if (bit_value) {
+			/* We know we have more than one block was allocated so
+			 * we need to find the size
+			 */
+			size_t bits_to_check =
+				heap->physical_blocks_allocators
+					[mem_block_iterator]->num_blocks - ptr_bit_array_position;
+
+			/* Neeeeeeeds optimization - thinking how to do it properly
+			 * each set bit in order after another means one allocated block.
+			 * When we get to 0 in such range we know that is last allocated block.
+			 * Testing bundles looks promising - need to investigate.
+			 */
+			for (i = ptr_bit_array_position;
+				i < bits_to_check + ptr_bit_array_position;
```

Comment:
Not realy - You sir just found a bug ;) It should be bits_to_check since we check ptr_bit_array_position before and it is not always buffer start position. So heap->physical_blocks_allocators[mem_block_iterator]->num_blocks could result in checking bits that are outside the bitmap/bitarray.